File: Rebex Components R6.11 - Samples/net-6.0/Ews/EwsBrowser_CS/Shared/RequestHandler.cs
//  
//   Rebex Sample Code License
// 
//   Copyright 2023, REBEX CR s.r.o.
//   All rights reserved.
//   https://www.rebex.net/
// 
//   Permission to use, copy, modify, and/or distribute this software for any
//   purpose with or without fee is hereby granted.
// 
//   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
//   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
//   OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
//   NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
//   HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
//   WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
//   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
//   OTHER DEALINGS IN THE SOFTWARE.
// 

using System;
using System.Collections;

using Rebex.Net;
using Rebex.Security.Certificates;

namespace Rebex.Samples
{
    /// <summary>
    /// Represents a handler for loading client certificates.
    /// </summary>
    public class RequestHandler : ICertificateRequestHandler
    {
        // certificate cache
        private static Hashtable _chosenCertificates = new Hashtable();

        public CertificateChain Request(TlsSocket socket, DistinguishedName[] issuers)
        {
            // try to locate certificate in cache
            string serverCertificateFingerprint = BitConverter.ToString(
                socket.ServerCertificate.LeafCertificate.GetCertHash());

            if (_chosenCertificates.Contains(serverCertificateFingerprint))
                return _chosenCertificates[serverCertificateFingerprint] as CertificateChain;

            // try to locate certificate in Store
            CertificateStore my = new CertificateStore("MY");
            Certificate[] certs;

            if (issuers.Length > 0)
            {
                certs = my.FindCertificates(
                    issuers,
                    CertificateFindOptions.IsTimeValid |
                    CertificateFindOptions.HasPrivateKey |
                    CertificateFindOptions.ClientAuthentication);
            }
            else
            {
                certs = my.FindCertificates(
                    CertificateFindOptions.IsTimeValid |
                    CertificateFindOptions.HasPrivateKey |
                    CertificateFindOptions.ClientAuthentication);
            }

            if (certs.Length == 0)
                return null;

            // confirm whether to use the certificate
            RequesetHandlerForm rhForm = new RequesetHandlerForm();
            rhForm.LoadData(certs);

            if (rhForm.ShowDialog() != System.Windows.Forms.DialogResult.OK)
                return null;

            CertificateChain chain = null;

            if (rhForm.Certificate != null)
                chain = CertificateChain.BuildFrom(rhForm.Certificate);

            // save chosen certificate to cache
            _chosenCertificates.Add(serverCertificateFingerprint, chain);

            return chain;
        }
    }
}

File: Rebex Components R6.11 - Samples/net-6.0/Ews/EwsBrowser_VB/Shared/RequestHandler.vb
'  
'   Rebex Sample Code License
' 
'   Copyright 2023, REBEX CR s.r.o.
'   All rights reserved.
'   https://www.rebex.net/
' 
'   Permission to use, copy, modify, and/or distribute this software for any
'   purpose with or without fee is hereby granted.
' 
'   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
'   EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
'   OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
'   NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
'   HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
'   WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
'   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
'   OTHER DEALINGS IN THE SOFTWARE.
' 

Imports System.Collections

Imports Rebex.Net
Imports Rebex.Security.Certificates


''' <summary>
''' Represents a handler for loading client certificates.
''' </summary>
Public Class RequestHandler
    Implements ICertificateRequestHandler
    ' certificate cache
    Private Shared _chosenCertificates As New Hashtable()

    Public Function Request(socket As TlsSocket, issuers As DistinguishedName()) As CertificateChain Implements ICertificateRequestHandler.Request
        ' try to locate certificate in cache
        Dim serverCertificateFingerprint As String = BitConverter.ToString(socket.ServerCertificate.LeafCertificate.GetCertHash())

        If _chosenCertificates.Contains(serverCertificateFingerprint) Then
            Return TryCast(_chosenCertificates(serverCertificateFingerprint), CertificateChain)
        End If

        ' try to locate certificate in Store
        Dim my As New CertificateStore("MY")
        Dim certs As Certificate()

        If issuers.Length > 0 Then
            certs = my.FindCertificates(issuers, CertificateFindOptions.IsTimeValid Or CertificateFindOptions.HasPrivateKey Or CertificateFindOptions.ClientAuthentication)
        Else
            certs = my.FindCertificates(CertificateFindOptions.IsTimeValid Or CertificateFindOptions.HasPrivateKey Or CertificateFindOptions.ClientAuthentication)
        End If

        If certs.Length = 0 Then
            Return Nothing
        End If

        ' confirm whether to use the certificate
        Dim rhForm As New RequesetHandlerForm()
        rhForm.LoadData(certs)

        If rhForm.ShowDialog() <> System.Windows.Forms.DialogResult.OK Then
            Return Nothing
        End If

        Dim chain As CertificateChain = Nothing

        If rhForm.Certificate IsNot Nothing Then
            chain = CertificateChain.BuildFrom(rhForm.Certificate)
        End If

        ' save chosen certificate to cache
        _chosenCertificates.Add(serverCertificateFingerprint, chain)

        Return chain
    End Function
End Class

